src/main/java/com/insightfullogic/honest_profiler/core/profiles/lean/LeanNode.java

```diff
@@ -0,0 +1,287 @@
+     *
+     * @return the {@link FrameInfo} associated with this LeanNode, or <null> if this LeanNode represents a thread
+     */
+    public FrameInfo getFrame()
+    {
+        return frame;
+    }
+
+    /**
+     * Returns the {@link NumericInfo} for this LeanNode
+     *
+     * @return the {@link NumericInfo} for this LeanNode
+     */
+    public NumericInfo getData()
+    {
+        return data;
+    }
+
+    /**
+     * Returns the parent of this LeanNode
+     *
+     * @return the parent of this LeanNode
+     */
+    public LeanNode getParent()
+    {
+        return parent;
+    }
+
+    /**
+     * Returns a {@link Collection} containing the children of this LeanNode.
+     *
+     * @return a {@link Collection} containing the children of this LeanNode
+     */
+    public Collection<LeanNode> getChildren()
+    {
+        return childMap.values();
+    }
+
+    /**
+     * Returns a boolean indicating whether this node represents a thread and is a {@link LeanThreadNode}.
+     *
+     * @return a boolean indicating whether this node represents a thread and is a {@link LeanThreadNode}
+     */
+    public boolean isThreadNode()
+    {
+        return false;
+    }
+
+    // Aggregation Methods
+
+    /**
+     * Aggregates teh information from a child {@link StackFrame} into the children of this LeanNode, updating the total
+     * data for this LeanNode in the process.
+     *
+     * The add method is called from the LogCollector, on the parent with the next (potentially last) child from the
+     * trace.
+     *
+     * @param nanos the number of nanoseconds between the {@link TraceStart} preceding the stack trace sample and the
+     *            {@link TraceStart} following it
+     * @param child the child {@link FrameInfo} of the current LeanNode
+     * @param last a boolean indicating if the child is the last in the stack trace sample
+     * @return this node
+     */
+    public LeanNode add(long nanos, FrameInfo child, boolean last)
+    {
+        // Non-self add, which updates total time and sample count only.
+        data.add(nanos, false);
+
+        return childMap.compute(
+            child,
+            (k, v) ->
+            // Create a new LeanNode if no children have been recorded for this FrameInfo
+            v == null ? (last ?
+            // New child is the last in stack, use "Self constructor".
+                new LeanNode(child, nanos, this)
+                // New child is not the last in stack, use "Non-self constructor".
+                : new LeanNode(child, this))
+                // Aggregate the information into an existing child
+                : (last ?
+                // Child is last in stack, use the "self add" on existing child
+                    v.addSelf(nanos) :
+                    // Child is not last in stack, return existing child (its "total" numbers will be updated when its
+                    // child is added)
+                    v));
```

Comment:
Yes, excellent catch ! This is indeed a lot simpler. As a bonus I can get rid of the "self constructor" as well, as well as a now obsolete constructor in NumericInfo.

Thanks !
Fixed in upcoming commit.